### PR TITLE
UI Overhaul

### PR DESCRIPTION
**Pull request by pajlow**
Hello, so this is my progress so far on #1150 . This PR gives an overview and is by far not the fully working installer. The next step would be to migrate all modals to svelte and then continue to handle all data exchanged between the components and the node backend properly. Just let me know if I should continue and if there are things you would handle differently. ATM the code is not very organized because I still have to implement a lot of things and clean up afterwards. After all dependencies are installed `npm run dev` is enough to get you started.

### CODE REVIEW

**Review by NeoTheThird — CHANGES_REQUESTED** [review]
I don't think we need `public/favicon.png`, do we?

File: public/index.html
Comment (NeoTheThird) [inline public/index.html:9]: remove and delete corresponding file
```suggestion
```

**Comment by pajlow** [comment]
No, it is basically just leftovers from the official svelte template.

File: rollup.config.js
Comment (NeoTheThird) [inline rollup.config.js:32]: we usually do two spaces indentation

File: src/App.svelte
Comment (NeoTheThird) [inline src/App.svelte:2]: indentation

File: src/main.js
Comment (NeoTheThird) [inline src/main.js:562]:
```suggestion
```

File: src/main.js
Comment (NeoTheThird) [inline src/main.js:574]:
```suggestion
  mainWindow.on("closed", function() {
    mainWindow = null;
```

**Review by NeoTheThird — CHANGES_REQUESTED** [review]
Very cool. Here are some things that jumped at me right away, but definitely on a good path, great work!

The duplicated files are due to the remaining pug components, i assume?

You might want to run `npm run lint-fix` in your repo to enforce the code style.

**Comment by NeoTheThird** [comment]
As for the conflicts, i would wait for #1553 to be merged and then rebase this branch on master. The gitignore conflict can be resolved by accepting both changes. Resolving `package.json` will need some brains to ensure consistency in the scripts and dependencies.  After package.json is resolved, `package-lock.json` can finally be deleted, re-built using `npm install`, and re-committed.

**Comment by NeoTheThird** [comment]
I sent you a PR to resolve the conflicts, if that's ok with you. I hope i did not break any of your changes. https://github.com/pajlow/ubports-installer/pull/1

Great work so far! I like the new state-aware way of handling modals and views. Still some stuff to be done, i think, but it looks to me as if this is on a very good way. How do you want to handle this further, are you planning to continue to work on it? Or do you want someone else to take over? What's your strategy regarding CSS? I'm also not afraid of properly changing how some stuff in the UI works, so if you feel like it's too difficult to replicate the old behavior in some parts, i'm open for new ideas.

**Comment by pajlow** [comment]
That is fine. I will merge it today. I just saw you deleted one development related command/dependency which I just for auto reload in svelte. 
I think I can get a lot of stuff done over the weekend and next week. By css you mean Bootstrap and overall styling? Bootstrap is bundled in theme.css inside the public/build folder and there is also one global css file for some things. Everything else I would handle inside the components.
Thinks I can't promise that they will work are some of the animations. I plan to get everything running and try to implement them afterwards.

**Comment by pajlow** [comment]
@NeoTheThird What are the advantages of changing _svelte.js_ and _rollup.config.js_ to _.mjs_ files? They broke some things for me when including _require_.  Do you have another solution for implementing reload when developing?

**Comment by NeoTheThird** [comment]
@pajlow you were using es6 import in those files, so they need to be modules. If you want to use commonjs require, feel free to change it back.

**Comment by pajlow** [comment]
@NeoTheThird is unlock modal the follow up to oem-lock? I have unfortunately no locked device to test it...

**Comment by NeoTheThird** [comment]
> @NeoTheThird is unlock modal the follow up to oem-lock? I have unfortunately no locked device to test it...

@pajlow no, it's used to display instructions before the os selection screen. the old canonical devices, for example, need to not be running android, before the installer can do anything. you can just select the bq aquaris e4.5 in the device selection to see what this looks like. btw, if you want to bring up any of the modals in the pug version, start the installer with `--debug` to bring up the debugger and enter `modals.show("oem-lock")` in the web console to see the modals.

**Comment by pajlow** [comment]
@NeoTheThird  I am almost done with the views. Next up will be the remaining modals but I have to rewrite some stuff because there is a lot of dom element creation going on there.

**Comment by NeoTheThird** [comment]
Awesome! So happy to finally get rid of the dom creation, that should be so much simpler now!

Would you mind if we already merge #1576? That one touches a lot of the backend files and also a few frontend things, but I'm happy to help resolve the conflicts, of course.

**Comment by pajlow** [comment]
@NeoTheThird No problem. I will merge it this evening after you open the PR.

**Comment by NeoTheThird** [comment]
alright, all done, won't touch the pug files from now on :) rebase away

**Comment by pajlow** [comment]
@NeoTheThird  Well, I really dont know what I did there. The local rebase looked fine. The merge messed things up. Now I got my commits twice. I got all of your changes except for the recent ones... And I still have to finish 3 modals.

**Comment by NeoTheThird** [comment]
Hmm, no idea how this happened. Don't worry, you're doing great, we'll find a way. Try pulling and merging master again, since there's a conflict, it might be github is just not showing the overall diff properly. don't worry about the messy commit history, we'll probably end up squashing this PR into a single commit anyways.

We could also do a "poor mans rebase", that is creating a new branch based off the latest master, copying all relevant changes over, creating a new commit and using that as a new PR. Since most your changes are replacing old files with new files in a different location, it's an okay amount of work.

Let me know if you need any help. Thanks again, so hyped for this!

**Comment by pajlow** [comment]
I am unfortunately lacking time during week days. I hope I will get some work done next weekend. I was hoping to get this finished before New Year's Eve.

**Comment by pajlow** [comment]
@NeoTheThird I need your help with a code snippet inside _options.pug_.
`ipcRenderer.on("user:configure", (event, osInstructs) => {
      if (!optionsAdded) {
        optionsAdded = true;
        osInstructs.forEach(addOption);
      }
      views.show("working", "particles");
      footer.underText.set("Please Configure the installation");
      modals.show("options");
      $("#btn-options-close").click(() => {
        setTimeout(() => ipcRenderer.send('install'), 250);
      });
    });`
The forEach-loop that runs for every OS instruction calls _addOption_ but the method has an option parameter. Where is the parameter provided? Is it provided by _osInstructs_? I looked in the backend inside _core.js_ but couldn't find the right property.

**Comment by NeoTheThird** [comment]
@pajlow sorry about the late reply. This is a bit of js syntax that allows you to do functional programming. Arrays have a [`forEach`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/forEach) function that accepts a function as a parameter. The function you pass to is then called once with every value. You'll often see a function literal expression used in these cases (like `someArray.forEach(item => console.log(item));`), but you can also just pass a reference to an existing function (like `someArray.forEach(console.log);`), which is what we do here.

**Comment by NeoTheThird** [comment]
Hey @pajlow, apologies for my recent inactivity, i'd like to pick this back up now so your great work can see the light of day soon. Is there still something you want to work on here before i resume reviewing? To resolve the conflicts, you should be able to resolve `package.json` and `src/core/api.js` fairly easily. Then, just delete `package-lock.json` and generate a new one by running `npm install`.

**Comment by pajlow** [comment]
@NeoTheThird Hi, sure thing. I would like to fix one minor error in the options modal and I did some refactoring I haven't pushed yet. I will look into it today and merge in the newest upstream changes. I will push it this evening.

**Comment by NeoTheThird** [comment]
alright! Now to satisfy ci, run `npm run lint-fix`, change the expected return of `api.getDeviceSelects()` in `src/core/helpers/api.spec.js` to that object you're producing, and remove the `npm run prerender && ` from the `build` script in `package.json`, and we should have some green checks :)

File: src/main.js
Comment (NeoTheThird) [inline src/main.js:53]:
```suggestion
    "Welcome to the UBports Installer version " + packageInfo.version + "!"
```

this should not have been changed back and probably happened during conflict resolution. we tried to get rid of the globals a while back wherever possible (not everywhere yet, but here it works).

**Review by NeoTheThird — CHANGES_REQUESTED** [review]
Hey, pretty cool what you've put together here, just looking through again. Not sure if you were finished yet, and i don't mean to be pushy or anything. Please, take the time you need :)

Just some things i've noticed

**style**

- [ ] many of the colors and text positions are a little off. it's ok if the installer looks different after the ui overhaul, but for release we'll have to polish some more
- [ ] the yumi animation no longer works (ie having the robot move from the footer to the device, to the spinning circle, etc)
- [ ] the waiting dots in the footer no longer work

**cleanup**

- [ ] `public/favicon.png` needs to go
- [x] remove `src/html` from `.gitignore`
- [x] code style is a lil' messy, we'll need to do a good linting of the svelte files as well. For one, we usually do two spaces of indentation and a trailing newline on every text file.
- [ ] two files remain in `src/pug/scripts`, they should be moved or integrated with other files
- [ ] your `package-lock.json` is generated by an old version of npm. Run `npm install -g npm && rm package-lock.json && npm i` to fix that. This is also why the loc on this pr is so high ^^

**functionality**

- [ ] the udev rules modal seems to come up every time (?)

now, of course you do not have to do all of these by yourself. it is amazing what you have already done here, let's find a way forward to enable people to actually use your great work. let me know how you would like to proceed. do you want me to tell you about things and then take a swing at implementing them yourself? or should i file a pr against your branch to get you started with some stuffs?

**Comment by NeoTheThird** [comment]
Alright, i sent you a PR here https://github.com/pajlow/ubports-installer/pull/2. That should handle most of the cleanup. make sure you read my note on the linter. you should be able to just rebase-and-merge that pr for the commits to show up here and then pull (use `git pull --rebase` to be sure, if you have local commits) your local branch to have the changes.

File: package.json
Comment (NeoTheThird) [inline package.json:31]:
```suggestion
    "docs": "npx jsdoc -c jsdoc-config.json",
    "diagram": "npx arkit -f src/main.js",
```

**Comment by NeoTheThird** [comment]
Don't forget to run `npm run lint` (and commit the result) once you're ready to have the svelte files styled. How much time are you calculating on the remaining issues? Thinking if we should do another release with small backend fixes before this, or if we should wait to avoid further conflicts. No pressure either way, tell me what's best for you.

**Comment by pajlow** [comment]
The linting is working perfectly, thanks! So far I resolved the udev modal issue and added Yumi. I think I will still need some time to go through the files and adjust the styling. Just go ahead with another release. I'm okay with resolving some conflicts.

**Comment by pajlow** [comment]
Remaining things are animations(particles, push files and download), accordion in dev mode modal (I used cards, because there is no component based solution for bootstrap 5 in Svelte yet). I can't check all modals because I can't mock all different scenarios. Is the overall installation process working? I have only one device I use as a daily driver.

**Comment by NeoTheThird** [comment]
Hey @pajlow, thanks again for your work, and apologies again for the long pause. I would now like to move forward with this. I think most of your work here is done, if you want to, i can take it from here. To allow me to merge this PR, you would have to accept and merge https://github.com/pajlow/ubports-installer/pull/3. If you don't have time for that right now, that's also fine, then i will merge your contributions into our master directly. Whatever works better for you :)

**Comment by pajlow** [comment]
Hi, welcome back! I am okay with you taking over. Did the current state of my repo work for you? I had some trouble with the  _remote.getGlobal__ calls.
I would like to continue to contribute after the merge. Further things to consider are the now stable bootstrap 5 and Sveltestrap which offers bootstrap elements as svelte components. :)

**Comment by NeoTheThird** [comment]
Thanks! No, it did not work completely, but i managed to get through in the dev tools with some trickery. But that's fine for now. My suggestion would be to merge this as-is and break some stuff, then incrementally fix things until everything works again and shoot for an 0.9 series release soon-ish. It's a good idea to try and move fast here, i think, because these constant conflict resolutions eat up too many resources. And if we need a hotfix on one of the 0.8 series releases, we can still do that by branching off an older state, we don't need to keep master pristine for that.

Alright. Do you maybe just want to merge https://github.com/pajlow/ubports-installer/pull/3, then i can use your PR here, that way your name is on the PR, too, and not just on the commits.

>  I would like to continue to contribute after the merge. Further things to consider are the now stable bootstrap 5 and Sveltestrap which offers bootstrap elements as svelte components. :)

Sounds great, looking forward to that!

**Comment by pajlow** [comment]
Merged!

**Review by NeoTheThird — APPROVED** [review]
:rocket:

**Comment by NeoTheThird** [comment]
I am *very* happy with this! After some fixing and intensive testing, i think we are very close to having everything working again. Great work! Really looking forward to doing more with this. This is going to unblock so many UI-related things! :tada: :tada: :tada:

**Comment by pajlow** [comment]
@capsia37 maybe you should open a new issue to get some thoughts on your designs. :)

**Pull request merged**
